cmd/titus-storage/ebs.go

```diff
@@ -0,0 +1,312 @@
+	// comes up, see it is attached, and mounts. In this case we need to still panic here, but it is "ok"
+	// In that situation we should not continue to detach a volume out from underneath it
+	l.Printf("Waiting up to %d seconds for %s to be not in use...", 30, actualDeviceName)
+	err = waitForDeviceToBeNotInUse(actualDeviceName, 30)
+	if err != nil {
+		return fmt.Errorf("Not detaching volume: %s", err)
+	}
+
+	// By the time we get here, we are confident the volume is attached, not in use, and safe to detatch
+	err = detatchEBS(ctx, c.ebsVolumeID, ec2Client, c.instanceID)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func getEc2Client() *ec2.EC2 {
+	sess := session.Must(session.NewSessionWithOptions(session.Options{
+		SharedConfigState: session.SharedConfigEnable,
+	}))
+	creds := credentials.NewCredentials(
+		&ec2rolecreds.EC2RoleProvider{
+			Client:       ec2metadata.New(sess),
+			ExpiryWindow: 5 * time.Minute,
+		},
+	)
+	sess.Config.Credentials = creds
+	*sess.Config.Region = os.Getenv("EC2_REGION")
+	return ec2.New(sess)
+}
+
+func waitForVolumeToBeUnused(ctx context.Context, ebsVolumeID string, ec2Client *ec2.EC2, timeout int) error {
+	l := logger.GetLogger(ctx)
+	for attempt := 0; attempt < timeout; attempt++ {
+		req := ec2.DescribeVolumeStatusInput{
+			VolumeIds: []*string{aws.String(ebsVolumeID)},
+		}
+		response, err := ec2Client.DescribeVolumeStatus(&req)
+		if err != nil {
+			return err
+		}
+		if len(response.VolumeStatuses) > 0 {
+			s := response.VolumeStatuses[0].VolumeStatus.Status
+			if *s == "ok" {
+				return nil
+			}
+			l.Printf("Volume Status %+v", *s)
+		}
+		time.Sleep(time.Duration(attempt) * time.Second)
+	}
+	return fmt.Errorf("Waited too long for volume to be unused")
+}
+
+func randDriveNamePicker() (string, error) {
+	ctr := 0
+	deviceName := "/dev/xvd"
+	runes := []rune("fghijklmnopqrstuvwxyz")
+	for {
+		if ctr >= len(runes) {
+			return "", fmt.Errorf("Ran out of drive names")
+		}
+		if !doesDriveExist(deviceName + string(runes[ctr])) {
+			break
+		}
+		ctr++
+	}
+	return deviceName + string(runes[ctr]), nil
+}
+
+func doesDriveExist(driveName string) bool {
+	if _, err := os.Stat(driveName); os.IsNotExist(err) {
+		for _, file := range listNvmeBlockDevices() {
+			if d, _ := ebsnvme.ScanDevice(file); d.Name == driveName {
+				return true
+			}
+		}
+		return false
+	}
+	return true
+}
+
+// waitForDeviceToDisappear is an inexpensive way to poll for an EBS
+// detachment without calling the EC2 API. It does this by waiting
+// for the block device it was attached with to dissapear from /dev
+func waitForDeviceToDisappear(ctx context.Context, driveName string, maxAttempts int) error {
+	l := logger.GetLogger(ctx)
+	var attempts int
+	for doesDriveExist(driveName) {
+		time.Sleep(2 * time.Second)
+		attempts++
+		if attempts >= maxAttempts {
+			return fmt.Errorf("waited the max %d seconds for %s to go, but it is still here", attempts*2, driveName)
```

Comment:
1. Use WithField
2. Record the start as time.Now(), and use time.since

Reply:
1. which field would you like added, the device name?
`l` should already have the task id and volume id

2. will do